File: blgdata1/public/js/admin_commission_adjustments.js
(function ($, api_url, Vue, swal, axios, location, undefined) {

    const client = commissionEngine.createAccessClient();
    commissionEngine.setupAccessTokenJQueryAjax();

    const vm = new Vue({
        el: ".tool-container",
        data: {
            dt: null,

            commissionTypes: [],
            commissionTypeState: "loaded", // loaded/fetching/error
            commissionType: "",

            commissionPeriods: [],
            commissionPeriodState: "loaded", // loaded/fetching/error
            commissionPeriodIndex: "",
            commissionPeriodView: "",

            purchaserId: null,
            userId: null,
            orderId: null,
            itemId: null,
            amount: null,
            level: null,
            remarks: null,
            autocompleteUrl: `${api_url}common/autocomplete/affiliates`,

            isCreateMode: 0,
            isEditMode: 0,
            isViewMode: 0,
            isProcessing: 0,


            filters: {
                id: null,
                member_id: null,
                commission_type_id: null,
                commission_period_id: null,
                purchaser_id: null,
                order_id: null,
                item_id: null,
                amount: null,
                level: null,
                remarks: null,
            },
            
            error: {
                message: null,
                type: null,
            },
        },
        mounted() {
            this.getCommissionTypes();
            this.initializeDataTables();
            this.initializeJQueryEvents();
            
        },
        methods: {
            initializeDataTables() {
                let _this = this;
                this.dt = $("#table-adjustments").DataTable({
                    processing: true,
                    serverSide: true,
                    responsive: true,
                    ajax: {
                        url: `${api_url}admin/commission-adjustments`
                    },
                    order: [[0, 'asc']],
                    columns: [
                        {
                            data: 'name',
                            render: function (data, type, row, meta) {
                                let member_id = row.member_id;
                                let member = row.name;
                                return `${member_id}: ${member}`;
                            }
                        },
                        {data: 'commission_type'},
                        {data: 'commission_period', className: "text-center"},
                        {data: 'amount', className: "text-center"},
                        {
                            data: 'actions',
                            className: "text-center",
                            render: function ( data, type, full, meta ) {
                                disabled_btn = '';
                                if(parseInt(full.is_locked) == 1){
                                    disabled_btn = 'disabled="disabled"';
                                }
                                
                                let $btnView = '<button class="btn btn-info btn-sm btn-view">VIEW</button> ';
                                let $btnEdit = '<button class="btn btn-warning btn-sm btn-edit" ' + disabled_btn + '">EDIT</button> ';
                                let $btnDelete = '<button class="btn btn-danger btn-sm btn-delete" ' + disabled_btn + '">DELETE</button> ';

                                let $options = $("<span>")
                                    .addClass("text-center")
                                    .append($btnView)
                                    .append($btnEdit)
                                    .append($btnDelete);

                                return $options[0].outerHTML;
                            }
                        },
                    ],
                    columnDefs: [
                        {responsivePriority: 1, targets: 0},
                        {responsivePriority: 2, targets: 1},
                        {responsivePriority: 3, targets: 2},
                    ]
                });
            },
            initializeJQueryEvents() {
                let _this = this;

                $('#table-adjustments tbody').on('click', '.btn-view', function ()  {
                    let data = _this.dt.row($(this).parents('tr')).data();
                    _this.showViewModal(data);
                });
                $('#table-adjustments tbody').on('click', '.btn-edit', function ()  {
                    let data = _this.dt.row($(this).parents('tr')).data();
                    _this.showEditModal(data);
                });
                $('#table-adjustments tbody').on('click', '.btn-delete', function ()  {
                    let data = _this.dt.row($(this).parents('tr')).data();
                    _this.deleteAdjustment(data);
                });
            },
            getCommissionTypes() {

                if (this.commissionTypeState === "fetching") return;

                this.commissionTypeState = "fetching";
                this.commissionTypes = [];

                client.get("common/commission-types/active-cash-manual")
                    .then(response => {
                        this.commissionTypes = response.data;
                        this.commissionTypeState = "loaded";

                    })
                    .catch(error => {
                        this.commissionTypeState = "error";
                    })
            },
            getCommissionPeriods(callback) {

                if (this.commissionPeriodState === "fetching") return;

                this.commissionPeriodState = "fetching";
                this.commissionPeriods = [];
                this.commissionPeriodIndex = "";
                
                this.filters.commission_type_id = this.commissionType;

                client.get(`common/commission-types/${this.filters.commission_type_id}/open-periods`)
                    .then(response => {
                        this.commissionPeriods = response.data;
                        this.commissionPeriodState = "loaded";
                        typeof callback == "function" && callback();
                    })
                    .catch(error => {
                        this.commissionPeriodState = "error";
                    })

            },
            showCreateModal() {

                this.isCreateMode = 1;
                this.isViewMode = 0;
                this.isEditMode = 0;
                this.filters.member_id = "";
                this.filters.purchaser_id = "";
                this.filters.order_id = "";
                this.filters.item_id = "";
                this.filters.amount = "";
                this.filters.level = "";
                this.filters.remarks = "";
                this.$refs.autocompletePurchaser.setDisabled(false);
                this.$refs.autocompleteMember.setDisabled(false);
                $('#modal-create').modal({backdrop: 'static', keyboard: false});

            },
            showViewModal(adjustment) {

                this.isViewMode = 1;
                this.isEditMode = 0;
                this.isCreateMode = 0;
                this.commissionType = adjustment.commission_type_id;
                this.getCommissionPeriods(() => {
                    this.commissionPeriodIndex = adjustment.commission_period_id;
                });
                this.filters.order_id = adjustment.transaction_id;
                this.filters.item_id = adjustment.item_id;
                this.filters.amount = adjustment.amount;
                this.filters.level = adjustment.level;
                this.filters.remarks = adjustment.remarks;
                this.$refs.autocompleteMember.setValue(`#${adjustment.member_id}: ${adjustment.name}`, adjustment.member_id);
                this.$refs.autocompleteMember.setDisabled(true);
                this.$refs.autocompletePurchaser.setValue(`#${adjustment.purchaser_id}: ${adjustment.purchaser_name}`, adjustment.purchaser_id);
                this.$refs.autocompletePurchaser.setDisabled(true);
                $('#modal-create').modal({backdrop: 'static', keyboard: false});
                
            },
            showEditModal(adjustment) {
                
                this.isEditMode = 1;
                this.isViewMode = 0;
                this.isCreateMode = 0;
                this.filters.id = adjustment.id;
                this.commissionType = adjustment.commission_type_id;
                this.getCommissionPeriods(() => {
                    this.commissionPeriodIndex = adjustment.commission_period_id;
                });
                this.filters.order_id = adjustment.transaction_id;
                this.filters.item_id = adjustment.item_id;
                this.filters.amount = adjustment.amount;
                this.filters.level = adjustment.level;
                this.filters.remarks = adjustment.remarks;
                this.$refs.autocompleteMember.setValue(`#${adjustment.member_id}: ${adjustment.name}`, adjustment.member_id);
                this.$refs.autocompleteMember.setDisabled(false);
                this.$refs.autocompletePurchaser.setValue(`#${adjustment.purchaser_id}: ${adjustment.purchaser_name}`, adjustment.purchaser_id);
                this.$refs.autocompletePurchaser.setDisabled(false);
                $('#modal-create').modal({backdrop: 'static', keyboard: false});
                
            },
            saveAdjustment() {
                
                let commissionPeriod = this.commissionPeriods[this.commissionPeriodIndex - 1];
                // this.filters.commission_period_id = commissionPeriod.id;
                this.filters.commission_period_id = this.commissionPeriodIndex;

                if(!this.filters.member_id) {
                    swal("Member is required.", "", "error");
                    return;
                }
                
                if(!this.filters.purchaser_id) {
                    swal("Purchaser is required.", "", "error");
                    return;
                }

                if(!this.filters.commission_period_id) {
                    swal("Period is required.", "", "error");
                    return;
                }

                if(this.isProcessing) return;

                swal({
                    title: "Confirm Commission Adjustments",
                    type: "warning",
                    confirmButtonClass: "btn-success",
                    confirmButtonText: "Confirm",
                    cancelButtonText: "Cancel",
                    showCancelButton: true,
                    closeOnConfirm: false,
                    showLoaderOnConfirm: true,
                }, () => {

                    this.isProcessing = 1;

                    client.post("admin/commission-adjustments/", this.filters).then(response => {
                        this.error.message = null;
                        this.error.type = null;

                        console.log(response.data);

                        this.dt.draw();
                        $('#modal-create').modal('hide');
                        swal('Success','','success');

                    }).catch(this.axiosErrorHandler).finally(()=> {
                        this.isProcessing = 0;
                    });

                });
            },
            updateAdjustment() {
                
                let commissionPeriod = this.commissionPeriods[this.commissionPeriodIndex - 1];
                this.filters.commission_period_id = this.commissionPeriodIndex;

                if(!this.filters.member_id) {
                    swal("Member is required.", "", "error");
                    return;
                }
                
                if(!this.filters.purchaser_id) {
                    swal("Purchaser is required.", "", "error");
                    return;
                }

                if(!this.filters.commission_period_id) {
                    swal("Period is required.", "", "error");
                    return;
                }

                if(this.isProcessing) return;

                swal({
                    title: "Confirm Updates for Commission Adjustments",
                    type: "warning",
                    confirmButtonClass: "btn-success",
                    confirmButtonText: "Confirm",
                    cancelButtonText: "Cancel",
                    showCancelButton: true,
                    closeOnConfirm: false,
                    showLoaderOnConfirm: true,
                }, () => {

                    this.isProcessing = 1;

                    client.post("admin/commission-adjustments/update", this.filters).then(response => {
                        this.error.message = null;
                        this.error.type = null;

                        console.log(response.data);

                        this.dt.draw();
                        $('#modal-create').modal('hide');
                        swal('Success','','success');

                    }).catch(this.axiosErrorHandler).finally(()=> {
                        this.isProcessing = 0;
                    });

                });
            },
            deleteAdjustment(data) {

                if(this.isProcessing) return;

                swal({
                    title: "Delete commission adjustments",
                    text: `Are you sure you want to delete the commission adjusments of ${data.name} (ID ${data.member_id})?`,
                    type: "warning",
                    confirmButtonClass: "btn-success",
                    confirmButtonText: "Confirm",
                    cancelButtonText: "Cancel",
                    showCancelButton: true,
                    closeOnConfirm: false,
                    showLoaderOnConfirm: true,
                }, () => {

                    this.isProcessing = 1;

                    client.post(`admin/commission-adjustments/${data.id}/delete`, this.filters).then(response => {
                        this.error.message = null;
                        this.error.type = null;

                        this.dt.draw();
                        swal('Success','','success');

                    }).catch(this.axiosErrorHandler).finally(()=> {
                        this.isProcessing = 0;
                    });

                });
            },
            axiosErrorHandler(error) {
                this.isProcessing = 0;

                let parse = commissionEngine.parseAxiosErrorData(error.response.data)

                this.error = parse.message;

                swal(this.error, "", 'error');
            },
        },
        watch: {
            commissionType: function (newType, oldType) {

                if (!!newType) {
                    this.commissionType = newType;
                    this.getCommissionPeriods();
                } else {
                    this.commissionTypes = [];
                    this.commissionType = "";
                    this.commissionPeriods = [];
                    this.commissionPeriodIndex = "";
                }
            }
        },
    });

}(jQuery, window.commissionEngine.API_URL, Vue, swal, axios, window.location));

File: dfodata/pages/admin_commission_adjustments.pl
print <<EOS;

<link rel="stylesheet" href="$commission_engine_api_url/css/app.css?v=$app_css_version" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.7/css/select2.min.css" />
<link rel="stylesheet" href="$commission_engine_api_url/css/select2-bootstrap.css"/>
<link rel="stylesheet" href="https://cdn.datatables.net/responsive/2.2.3/css/responsive.bootstrap.min.css">
<link rel="stylesheet" href="$commission_engine_api_url/css/admin_commission_adjustments.css?v=1.0&app=$app_css_version" />

<div class="tool-container tool-container--default" v-cloak>

	<div class="row">
        <div class="col-md-12">
            <h4 class="admin-money-title">Commission Adjustments</h4>
        </div>
    </div>
    <div class="panel panel-default">

        <div class="panel-body">

			<div class="row margin-top-bottom">
				<div class="col-md-3">
					<button type="button" class="btn btn-block btn-success" v-on:click.prevent="showCreateModal">
						New adjustment
					</button>
				</div>
				<div class="col-md-3"></div>
				<div class="col-md-3"></div>
				<div class="col-md-3"></div>
			</div>

			<div class="row">
				<div class="col-md-12">
					<table class="table table-striped table-bordered dt-responsive nowrap" id="table-adjustments" style="width:100%">
						<thead class="table__header table__header--bg-primary">
							<tr class="table__row">
								<td class="table__cell">Name</td>
								<td class="table__cell">Commission Type</td>
								<td class="table__cell">Commission Period</td>
								<td class="table__cell">Amount</td>
								<td class="table__cell"></td>
							</tr>
						</thead>
		                <tbody class="table__body">
						</tbody>
					</table>
				</div>
			</div>

		</div>
	</div>

	<!-- NEW ADJUSTMENTS MODAL -->
    <div class="modal fade" id="modal-create" role="dialog" aria-labelledby=modal-create-title">
        <div class="modal-dialog modal-md" role="document">
            <div class="modal-content">
                <div class="modal-header" style="display: block;">
                    <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
                    <h4 v-if="isEditMode === 1" class="modal-title" id="modal-create-title">
                        <i class="fa fa-plus" aria-hidden="true"></i> Edit adjustment
                    </h4>
					<h4 v-else-if="isViewMode === 1" class="modal-title" id="modal-create-title">
                        <i class="fa fa-plus" aria-hidden="true"></i> View adjustment
                    </h4>
					<h4 v-else class="modal-title" id="modal-create-title">
                        <i class="fa fa-plus" aria-hidden="true"></i> New adjustment
                    </h4>
                </div>
                <div class="modal-body">
            		<form class="form-horizontal" style="margin-bottom: 15px;">
						<div class="form-group">
							<div class="col-md-8">
								<label class="control-label" for="userId">Step 1. Select member</label>
								<select2-autocomplete-member ref="autocompleteMember" v-bind:url="autocompleteUrl" v-model="filters.member_id"></select2-autocomplete-member>
							</div>
						</div>
						<div class="form-group">
							<div class="col-md-12">
								<label for="commission_type">Commission Type</label>
								<select name="commission_type" id="commission_type" class="form-control" v-model="commissionType" :disabled="!!isViewMode">
									<option value="" selected disabled>Select a type</option>
									<option v-for="(type, index) in commissionTypes"
											v-bind:value="type.id"
											v-bind:key="type.id">
										{{ type.name }}
									</option>
								</select>
							</div>
						</div>
						<div class="form-group">
							<div class="col-md-12">
								<label for="commission_period">Commission Period <span class="text-danger">*</span></label>
								<select disabled name="commission_period" id="commission_period" class="form-control" v-model="commissionPeriodIndex" v-bind:disabled="commissionPeriodState !== 'loaded'" :disabled="!!isViewMode">
									<option v-if="commissionPeriodState === 'fetching'" value="" selected disabled>
										Fetching...
									</option>
									<option v-else-if="commissionPeriodState === 'error'" value="" selected disabled>
										Error
									</option>
									<option v-else-if="commissionPeriodState === 'loaded'" value="" selected disabled>
										Select a commission period
									</option>
									<option v-for="(period, index) in commissionPeriods"
											v-bind:value="period.id" 
											v-bind:key="index">
										{{ period.start_date }} to {{ period.end_date }}
									</option>
								</select>
								<a style="display: none;" v-show="commissionPeriodState === 'error'" v-on:click.prevent="getCommissionPeriods" class="help-block text-danger">Unable to fetch. Click here to try again.</a>
							</div>
						</div>
						<div class="form-group">
							<div class="col-md-12">
								<label class="control-label" for="commission_period"> Step 4. Input transaction details</label>
							</div>
							<div class="col-md-8">
								<label class="control-label" for="purchaserId">Purchaser</label>
								<select2-autocomplete-member ref="autocompletePurchaser" v-bind:url="autocompleteUrl" v-model="filters.purchaser_id"></select2-autocomplete-member>
							</div>
						</div>
						<div class="form-group">
							<div class="col-sm-3">
								<label class="control-label label-light" for="">Order ID</label>
								<input data-validation="required" type="number" class="form-control" v-model="filters.order_id" id="form-order_id" placeholder="" :disabled="!!isViewMode">
							</div>
							<div class="col-sm-3">
								<label class="control-label label-light" for="">Item ID</label>
								<input data-validation="required" type="number" class="form-control" v-model="filters.item_id" id="form-item_id" placeholder="" :disabled="!!isViewMode">
							</div>
							<div class="col-sm-3">
								<label class="control-label label-light" for="">Amount</label>
								<input data-validation="required" type="number" class="form-control" v-model="filters.amount" id="form-amount" placeholder="" :disabled="!!isViewMode">
							</div>
							<div class="col-sm-3">
								<label class="control-label label-light" for="">Level</label>
								<input data-validation="required" type="number" class="form-control" v-model="filters.level" id="form-level" placeholder="" :disabled="!!isViewMode">
							</div>
						</div>
						<div class="form-group">
							<div class="col-md-12">
								<label class="control-label" for="">Step 5: Remarks</label>
							</div>
							<div class="col-md-12">
								<textarea data-validation="required" class="form-control" rows="3" v-model="filters.remarks" id="form-remarks" :disabled="!!isViewMode"></textarea>
							</div>
						</div>
            		</form>
                </div>
                <div class="modal-footer">
					<button v-if="isCreateMode === 1" type="submit" class="btn btn-primary" id="btn-set" v-bind:disabled="isProcessing === 1" v-on:click.prevent="saveAdjustment">Set</button>
					<button v-else-if="isEditMode === 1" type="submit" class="btn btn-primary" id="btn-set" v-bind:disabled="isProcessing === 1" v-on:click.prevent="updateAdjustment">Edit</button>
                    <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                </div>
            </div>
        </div>
    </div>

</div>


<script src="$commission_engine_api_url/js/app.js?v=$app_js_version"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.7/js/select2.full.min.js"></script>
<script src="https://cdn.datatables.net/responsive/2.2.3/js/dataTables.responsive.min.js"></script>
<script src="https://cdn.datatables.net/responsive/2.2.3/js/responsive.bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.24.0/moment.min.js" integrity="sha256-4iQZ6BVL4qNKlQ27TExEhBN1HFPvAvAMbFavKKosSWQ=" crossorigin="anonymous"></script>
<script src="$commission_engine_api_url/js/admin_commission_adjustments.js?v=2.0&app=$app_js_version"></script>


EOS
1;